include/circt/Dialect/HW/HWTypes.h

```diff
@@ -110,6 +110,30 @@ BaseTy type_dyn_cast(Type type) {
   return type_cast<BaseTy>(type);
 }
 
+/// Utility type that wraps a type that may be one of several possible Types.
+/// This is similar to std::variant but is implemented for mlir::Type, and it
+/// understands how to handle type aliases.
+template <typename... Types>
+class TypeVariant
+    : public ::mlir::Type::TypeBase<TypeVariant<Types...>, mlir::Type,
+                                    mlir::TypeStorage> {
+  using mlir::Type::TypeBase<TypeVariant<Types...>, mlir::Type,
+                             mlir::TypeStorage>::Base::Base;
+
+public:
+  // Support LLVM isa/cast/dyn_cast to one of the possible types.
+  static bool classof(Type other) { return type_isa<Types...>(other); }
+
+  // Support C++ explicit conversions to the underlying types. These are
+  // explicit because the compiler can't statically know which conversion is
+  // actually safe to use.
+  template <typename T,
+            std::enable_if_t<(std::is_same<T, Types>::value || ...)>>
```

Comment:
This is great, glad you got this working!

FWIW, I think this should be:
```suggestion
            typename = std::enable_if_t<(std::is_same<T, Types>::value || ...)>>
```
Confirmed by trying it out, as written I'm not sure this ever will be enabled/valid.

Also, FWIW, and sorry for not noticing before, but actually this operator is for normal (explicit) conversions, like this functional-style conversion (or `static_cast<IntegerType>()`, etc):
```
auto foo = IntegerType(myAddOp.getType());
```

And I don't think this operator is part of or necessary for LLVM's "Casting" support (`cast<>`, `isa<>`, `dyn_cast<>`, ...).
Indeed, everything compiles and tests pass if we drop it entirely (and maybe why it worked without the `typename =` bit).

However, having come this far we can still support conversions if we change this as suggested above.
Keeping it explicit still seems good to me so there are no surprises when compiler uses wrong type and the cast fails unexpectedly.

Reply:
Thanks for catching that!

> And I don't think this operator is part of or necessary for LLVM's "Casting" support (cast<>, isa<>, dyn_cast<>, ...).
Indeed, everything compiles and tests pass if we drop it entirely (and maybe why it worked without the typename = bit).

Right, I had thought about this when I originally wrote it. I had honestly started by copying the implementation of `TypeAliasOr` and modifying it to match my needs, and in `TypeAliasOr`, we have implicit conversions defined. For `TypeAliasOr`, it makes a bit more sense to have the implicit conversion because there's only one canonical type in the end.

For this new TypeVariant class, maybe we should just drop the explicit conversions until we have a use case for it. It's easy enough to add this code back in later if we end up needing it.

Reply:
Sounds good to me!